change-notes/1.21/analysis-csharp.md

```diff
@@ -9,10 +9,24 @@
 
 ## Changes to code extraction
 
-* Named attribute arguments are now extracted.
+* The following C# 8 features are now extracted:
+    - Range expressions
+    - Recursive patterns
 
 ## Changes to QL libraries
 
 * The class `Attribute` has two new predicates: `getConstructorArgument()` and `getNamedArgument()`. The first predicate returns arguments to the underlying constructor call and the latter returns named arguments for initializing fields and properties.
+* The following QL classes have been added to model C# 8 features:
+    - Class `IndexExpr` models from-end index expressions, for example `^1`
+    - Class `PatternExpr` is an `Expr` that appears in a pattern. It has the new subclasses `DiscardPatternExpr`, `LabeledPatternExpr`, `RecursivePatternExpr`, `TypeAccessPatternExpr`, `TypePatternExpr`, and `VariablePatternExpr`.
+    - Class `PatternMatch` models a pattern being matched. It has the subclasses `Case` and `IsExpr`.
+    - Class `PositionalPatternExpr` models position patterns, for example `(int x, int y)`
+    - Class `PropertyPatternExpr` models property patterns, for example `Length: int len`
+    - Class `RangeExpr` models range expressions, for example `1..^1`
+    - Class `SwitchCaseExpr` models the arm of a switch expression, for example `(false, false) => true`
+    - Class `SwitchExpr` models `switch` expressions, for example `(a, b) switch { ... }`
+    - Classes `IsConstantExpr`, `IsTypeExpr` and `IsPatternExpr` are deprecated in favour of `IsExpr`
```

Comment:
Oxford comma

Reply:
I would like to merge this PR as it has to be done simultaneously with an internal PR. @calumgrant : Could you make the change note update in a follow-up PR please?